frontend/src/pages/Player/Toolbar/DevToolsWindowV2/ConsolePage/ConsolePage.tsx

```diff
@@ -1,199 +1,147 @@
+					return false
 				}
-			})()
+
+				return logEdge.node.message
+					.toLocaleLowerCase()
+					.includes(filter.toLocaleLowerCase())
+			})
 		}
-	}, [session?.messages_url, session?.secure_id, session_secure_id])
 
-	useEffect(() => {
-		setParsedMessages(
-			data?.messages?.map((m: ConsoleMessage, i) => {
-				return {
-					...m,
-					id: i,
-				}
-			}) ?? [],
-		)
-	}, [data?.messages])
+		return data?.sessionLogs
+	}, [filter, data?.sessionLogs])
 
 	// Logic for scrolling to current entry.
 	useEffect(() => {
-		if (state !== ReplayerState.Playing) {
+		if (state !== ReplayerState.Playing || !autoScroll) {
 			return
 		}
-		if (parsedMessages?.length) {
-			let msgIndex = 0
+		if (messagesToRender?.length) {
+			let cursor = ''
 			let msgDiff: number = Number.MAX_VALUE
-			for (let i = 0; i < parsedMessages.length; i++) {
+			for (let i = 0; i < messagesToRender.length; i++) {
 				const currentDiff: number =
-					time - (parsedMessages[i].time - parsedMessages[0].time)
+					time -
+					(new Date(messagesToRender[i].node.timestamp).getTime() -
+						new Date(messagesToRender[0].node.timestamp).getTime())
 				if (currentDiff < 0) break
 				if (currentDiff < msgDiff) {
-					msgIndex = i
+					cursor = messagesToRender[i].cursor
 					msgDiff = currentDiff
 				}
 			}
-			setCurrentMessage(msgIndex)
+			setSelectedCursor(cursor)
 		}
-	}, [state, time, parsedMessages])
-
-	const messagesToRender = useMemo(() => {
-		const currentMessages = parsedMessages?.filter((m) => {
-			// if the console type is 'all', let all messages through. otherwise, filter.
-			if (logLevel === LogLevel.All) {
-				return true
-			} else if (m.type === logLevel.toLowerCase()) {
-				return true
-			}
-			return false
-		})
-
-		if (!currentMessages) {
-			return []
-		}
-
-		if (filter !== '') {
-			return currentMessages.filter((message) => {
-				if (!message.value) {
-					return false
-				}
-
-				switch (typeof message.value) {
-					case 'string':
-						return message.value
-							.toLocaleLowerCase()
-							.includes(filter.toLocaleLowerCase())
-					case 'object':
-						return message.value.some((line: string | null) => {
-							return line
-								?.toString()
-								.toLocaleLowerCase()
-								.includes(filter.toLocaleLowerCase())
-						})
-					default:
-						return false
-				}
-			})
-		}
-
-		return currentMessages.filter((message) => message?.value?.length)
-	}, [logLevel, filter, parsedMessages])
+	}, [state, time, messagesToRender, autoScroll])
 
 	const virtuoso = useRef<VirtuosoHandle>(null)
-	// eslint-disable-next-line react-hooks/exhaustive-deps
-	const scrollFunction = useCallback(
-		_.debounce((index: number) => {
-			if (virtuoso.current) {
+	useEffect(() => {
+		if (
+			isPlayerReady && // ensure Virtuoso component is actually rendered
+			virtuoso.current &&
+			messagesToRender
+		) {
+			const index = messagesToRender.findIndex(
+				(logEdge) => logEdge.cursor === selectedCursor,
+			)
```

Comment:
since this could be an expensive find, does it make sense to pull this into it's own `useMemo` value that just depends on `messagesToRender` and `selectedCursor` so that this doesn't change when the rest of the dependencies change?

Reply:
Thanks for pairing. I've adjusted to use `useMemo` in aa48158ff9d44978ae6941535b7a8080decc913f

Will do another test on the render preview when it deploys.